### ISSUE TITLE
Consider using runtime.entersyscall() instead of runtime.entersyscallblock()

### ISSUE THREAD

**Issue by fako1024** [fako1024/slimcap#72]
Currently, the optimized assembler logic for `PPOLL` utilizes `runtime.entersyscallblock()`. The rationale behind this was to allow the scheduler to release resources faster (assuming the poll will be blocking for a long time). However, practical tests / benchmarks have shown that while under heavy load / traffic (or to be more precise: when a lot of polling has to be done) the time spent blocking is actually very short (which defies the whole concept of indicating to the scheduler that this won't be the case) and apparently the "misinformation" to the scheduler causes quite some overhead (most likely because it has to reschedule right away instead of just keeping the current resources):
```
.../NextPayloadZeroCopy_NoBlock_10kiBx512-4  88184865  68.01 ns/op
.../NextPayloadZeroCopy_Block10kiBx512-4     30295363  176.7 ns/op
```
On the other hand, if there's little traffic (and hence the blocking state might be retained a while) we a) don't really care about squeezing out that last bit of performance and b) the scheduler will release the resources after a while (O(20µs)) anyway (achieving the same state `runtime.entersyscallblock()` essentially just "forces" right away).

**Comment by fako1024** [comment]
Some benchmarks (three different systems):
```
cpu: Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
                                                 │   sec/op    │    sec/op     vs base                │
CaptureMethods/NextPacket_10485kiBx4-4             53.58n ± 0%   53.81n ±  0%   +0.44% (p=0.000 n=50)
CaptureMethods/NextPacketInPlace_10485kiBx4-4      22.33n ± 0%   22.33n ±  0%        ~ (p=0.994 n=50)
CaptureMethods/NextPayload_10485kiBx4-4            47.34n ± 0%   47.41n ±  0%        ~ (p=0.255 n=50)
CaptureMethods/NextPayloadInPlace_10485kiBx4-4     18.67n ± 6%   19.79n ±  6%        ~ (p=0.191 n=50)
CaptureMethods/NextPayloadZeroCopy_10485kiBx4-4    11.16n ± 0%   11.16n ±  0%        ~ (p=0.725 n=50)
CaptureMethods/NextIPPacket_10485kiBx4-4           47.38n ± 0%   47.51n ±  0%   +0.28% (p=0.010 n=50)
CaptureMethods/NextIPPacketInPlace_10485kiBx4-4    19.57n ± 0%   19.51n ±  4%   -0.31% (p=0.027 n=50)
CaptureMethods/NextIPPacketZeroCopy_10485kiBx4-4   11.19n ± 0%   11.20n ±  0%        ~ (p=0.592 n=50)
CaptureMethods/NextPacketFn_10485kiBx4-4           12.86n ± 0%   12.88n ±  0%        ~ (p=0.257 n=50)
CaptureMethods/NextPacket_10kiBx512-4              132.4n ± 2%   140.8n ±  7%        ~ (p=0.276 n=50)
CaptureMethods/NextPacketInPlace_10kiBx512-4       165.1n ± 2%   144.9n ±  6%  -12.27% (p=0.000 n=50)
CaptureMethods/NextPayload_10kiBx512-4             150.7n ± 2%   140.9n ± 10%   -6.47% (p=0.016 n=50)
CaptureMethods/NextPayloadInPlace_10kiBx512-4      165.2n ± 3%   142.6n ±  4%  -13.65% (p=0.000 n=50)
CaptureMethods/NextPayloadZeroCopy_10kiBx512-4     167.4n ± 4%   143.3n ±  9%  -14.40% (p=0.000 n=50)
CaptureMethods/NextIPPacket_10kiBx512-4            151.4n ± 4%   142.7n ± 14%   -5.75% (p=0.013 n=50)
CaptureMethods/NextIPPacketInPlace_10kiBx512-4     162.4n ± 2%   144.4n ±  4%  -11.06% (p=0.000 n=50)
CaptureMethods/NextIPPacketZeroCopy_10kiBx512-4    171.5n ± 3%   142.2n ±  3%  -17.09% (p=0.000 n=50)
CaptureMethods/NextPacketFn_10kiBx512-4            167.1n ± 4%   144.8n ±  8%  -13.38% (p=0.000 n=50)
geomean                                            59.98n        57.12n         -4.77%
```
```
cpu: Intel(R) Celeron(R) N5105 @ 2.00GHz
                                                 │   sec/op    │   sec/op     vs base                │
CaptureMethods/NextPacket_10485kiBx4-4             79.39n ± 0%   79.80n ± 0%   +0.51% (p=0.000 n=50)
CaptureMethods/NextPacketInPlace_10485kiBx4-4      34.04n ± 1%   34.20n ± 0%        ~ (p=0.129 n=50)
CaptureMethods/NextPayload_10485kiBx4-4            68.41n ± 1%   68.80n ± 0%   +0.58% (p=0.027 n=50)
CaptureMethods/NextPayloadInPlace_10485kiBx4-4     29.77n ± 0%   29.96n ± 1%        ~ (p=0.060 n=50)
CaptureMethods/NextPayloadZeroCopy_10485kiBx4-4    15.91n ± 0%   15.91n ± 0%        ~ (p=0.809 n=50)
CaptureMethods/NextIPPacket_10485kiBx4-4           70.93n ± 0%   71.27n ± 0%   +0.47% (p=0.031 n=50)
CaptureMethods/NextIPPacketInPlace_10485kiBx4-4    31.86n ± 1%   31.69n ± 1%        ~ (p=0.893 n=50)
CaptureMethods/NextIPPacketZeroCopy_10485kiBx4-4   16.15n ± 0%   16.14n ± 0%        ~ (p=0.874 n=50)
CaptureMethods/NextPacketFn_10485kiBx4-4           21.73n ± 9%   20.14n ± 7%        ~ (p=0.083 n=50)
CaptureMethods/NextPacket_10kiBx512-4              129.5n ± 2%   121.5n ± 1%   -6.14% (p=0.000 n=50)
CaptureMethods/NextPacketInPlace_10kiBx512-4       122.0n ± 1%   112.5n ± 1%   -7.83% (n=50)
CaptureMethods/NextPayload_10kiBx512-4             122.6n ± 1%   119.5n ± 1%   -2.53% (p=0.000 n=50)
CaptureMethods/NextPayloadInPlace_10kiBx512-4      122.0n ± 1%   112.4n ± 1%   -7.87% (n=50)
CaptureMethods/NextPayloadZeroCopy_10kiBx512-4     126.8n ± 1%   112.6n ± 1%  -11.23% (n=50)
CaptureMethods/NextIPPacket_10kiBx512-4            122.9n ± 1%   119.2n ± 1%   -3.01% (p=0.000 n=50)
CaptureMethods/NextIPPacketInPlace_10kiBx512-4     122.0n ± 1%   112.3n ± 0%   -7.95% (n=50)
CaptureMethods/NextIPPacketZeroCopy_10kiBx512-4    127.3n ± 1%   112.7n ± 1%  -11.51% (n=50)
CaptureMethods/NextPacketFn_10kiBx512-4            124.9n ± 1%   112.1n ± 1%  -10.21% (n=50)
geomean                                            65.53n        62.78n        -4.19%
```
```
cpu: Intel(R) Core(TM) i3-2100T CPU @ 2.50GHz
                                                 │   sec/op    │   sec/op     vs base                │
CaptureMethods/NextPacket_10485kiBx4-4             103.3n ± 1%   103.1n ± 1%        ~ (p=0.218 n=50)
CaptureMethods/NextPacketInPlace_10485kiBx4-4      39.94n ± 1%   40.12n ± 1%        ~ (p=0.341 n=50)
CaptureMethods/NextPayload_10485kiBx4-4            88.76n ± 1%   88.37n ± 1%        ~ (p=0.084 n=50)
CaptureMethods/NextPayloadInPlace_10485kiBx4-4     32.90n ± 3%   32.55n ± 3%        ~ (p=0.907 n=50)
CaptureMethods/NextPayloadZeroCopy_10485kiBx4-4    16.38n ± 0%   16.32n ± 0%   -0.37% (p=0.003 n=50)
CaptureMethods/NextIPPacket_10485kiBx4-4           89.80n ± 1%   89.12n ± 1%        ~ (p=0.076 n=50)
CaptureMethods/NextIPPacketInPlace_10485kiBx4-4    33.51n ± 2%   33.30n ± 1%        ~ (p=0.759 n=50)
CaptureMethods/NextIPPacketZeroCopy_10485kiBx4-4   16.68n ± 0%   16.62n ± 0%   -0.33% (p=0.016 n=50)
CaptureMethods/NextPacketFn_10485kiBx4-4           19.52n ± 1%   19.43n ± 1%        ~ (p=0.200 n=50)
CaptureMethods/NextPacket_10kiBx512-4              214.0n ± 1%   236.0n ± 2%  +10.26% (p=0.000 n=50)
CaptureMethods/NextPacketInPlace_10kiBx512-4       197.1n ± 2%   162.7n ± 2%  -17.45% (n=50)
CaptureMethods/NextPayload_10kiBx512-4             182.6n ± 1%   230.0n ± 2%  +25.96% (p=0.000 n=50)
CaptureMethods/NextPayloadInPlace_10kiBx512-4      195.5n ± 1%   155.3n ± 1%  -20.54% (n=50)
CaptureMethods/NextPayloadZeroCopy_10kiBx512-4     189.8n ± 2%   150.2n ± 2%  -20.89% (n=50)
CaptureMethods/NextIPPacket_10kiBx512-4            180.1n ± 1%   224.8n ± 2%  +24.85% (p=0.000 n=50)
CaptureMethods/NextIPPacketInPlace_10kiBx512-4     192.9n ± 2%   155.0n ± 2%  -19.62% (n=50)
CaptureMethods/NextIPPacketZeroCopy_10kiBx512-4    177.3n ± 2%   150.0n ± 2%  -15.37% (n=50)
CaptureMethods/NextPacketFn_10kiBx512-4            198.8n ± 4%   154.8n ± 3%  -22.14% (n=50)
geomean                                            86.09n        82.45n        -4.24%

```

**Comment by fako1024** [comment]
After giving it some more thought I've decided to forego this. Rationale:

- The impact on the "default" use case of a sufficiently large buffer / block size is non-existent (as proven in the benchmarks), so we'd be talking about an edge case optimization right from the get-go.
- The idea of using `runtime.entersyscallblock()` was to allow releasing resources to the scheduler asap, so that _other_ goroutines / captures can benefit from that. That scenario is not covered in the benchmark, so it's impossible to foresee / measure any (likely) detrimental effects on e.g. systems with many interfaces (not so much an edge case anymore).
- The improvements for small buffers are substantial, but: In order to benefit, the buffer must be quite small (in this case it can only hold about 80 packets per block). Such case is just unrealistic / unnecessary (issue can easily be solved by increasing the buffer to a reasonable size).

By my math, assuming that the scheduler decides to release the resources on its own without the blocking hint after 20µs (unfortunately I can't find the original documentation anymore, but I did read that number) and that fetching a single packet from the current block takes about 10-20 ns, the point of diminishing returns would be of the order of 1000-2000 packets / block (which, at a snaplen of 128 bytes corresponds to 128-256 kiB per block). If performance is so critical that this matters, I'm sure the user can afford an overall ring buffer of 1MiB per interface.

@els0r FYI.
